# Cancer Detection

Breast Cancer Wisconsin (Diagnostic) Data Set

Data: https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/breast-cancer-wisconsin.data

In [1]:
import numpy as np
import pandas as pd

In [2]:
col = ['id', 'Clump Thickness', 'Uniformity of Cell Size', 
       'Uniformity of Cell Shape', 'Marginal Adhesion', 
       'Single Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin',
       'Normal Nucleoli', 'Mitoses', 'Class']

df = pd.read_csv("breast-cancer-wisconsin.data", names=col, header=None)
df.head()

id  Clump Thickness  Uniformity of Cell Size  \
0  1000025                5                        1   
1  1002945                5                        4   
2  1015425                3                        1   
3  1016277                6                        8   
4  1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

  Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0           1                3                1        1      2  
1          10                3                2        1      2  
2           2                3                1        1      2  
3           4                3                7        1      2  
4           1                3                1        1      2

# Data Pre-processing

In [3]:
# Check for null values
df.isnull().sum()

id                             0
Clump Thickness                0
Uniformity of Cell Size        0
Uniformity of Cell Shape       0
Marginal Adhesion              0
Single Epithelial Cell Size    0
Bare Nuclei                    0
Bland Chromatin                0
Normal Nucleoli                0
Mitoses                        0
Class                          0
dtype: int64

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699 entries, 0 to 698
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   id                           699 non-null    int64 
 1   Clump Thickness              699 non-null    int64 
 2   Uniformity of Cell Size      699 non-null    int64 
 3   Uniformity of Cell Shape     699 non-null    int64 
 4   Marginal Adhesion            699 non-null    int64 
 5   Single Epithelial Cell Size  699 non-null    int64 
 6   Bare Nuclei                  699 non-null    object
 7   Bland Chromatin              699 non-null    int64 
 8   Normal Nucleoli              699 non-null    int64 
 9   Mitoses                      699 non-null    int64 
 10  Class                        699 non-null    int64 
dtypes: int64(10), object(1)
memory usage: 60.2+ KB


Note: In the Bare Nuclei feature, note that the datatype is object, but its meant to be Integer.

In [5]:
df['Bare Nuclei'].describe()

count     699
unique     11
top         1
freq      402
Name: Bare Nuclei, dtype: object

In [6]:
df['Bare Nuclei'].value_counts()

1     402
10    132
5      30
2      30
3      28
8      21
4      19
?      16
9       9
7       8
6       4
Name: Bare Nuclei, dtype: int64

Note there is a `?` feature, instead of an integer.

In [7]:
# To check the rows that have "?" as a value
df[df['Bare Nuclei'] == "?"]

id  Clump Thickness  Uniformity of Cell Size  \
23   1057013                8                        4   
40   1096800                6                        6   
139  1183246                1                        1   
145  1184840                1                        1   
158  1193683                1                        1   
164  1197510                5                        1   
235  1241232                3                        1   
249   169356                3                        1   
275   432809                3                        1   
292   563649                8                        8   
294   606140                1                        1   
297    61634                5                        4   
315   704168                4                        6   
321   733639                3                        1   
411  1238464                1                        1   
617  1057067                1                        1   

     Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
23                          5                  1                            2   
40                          6                  9                            6   
139                         1                  1                            1   
145                         3                  1                            2   
158                         2                  1                            3   
164                         1                  1                            2   
235                         4                  1                            2   
249                         1                  1                            2   
275                         3                  1                            2   
292                         8                  1                            2   
294                         1                  1                            2   
297                         3                  1                            2   
315                         5                  6                            7   
321                         1                  1                            2   
411                         1                  1                            1   
617                         1                  1                            1   

    Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
23            ?                7                3        1      4  
40            ?                7                8        1      2  
139           ?                2                1        1      2  
145           ?                2                1        1      2  
158           ?                1                1        1      2  
164           ?                3                1        1      2  
235           ?                3                1        1      2  
249           ?                3                1        1      2  
275           ?                2                1        1      2  
292           ?                6               10        1      4  
294           ?                2                1        1      2  
297           ?                2                3        1      2  
315           ?                4                9        1      2  
321           ?                3                1        1      2  
411           ?                2                1        1      2  
617           ?                1                1        1      2

They are mostly of the Benign class "2"

In [8]:
len(df[df['Bare Nuclei'] == "?"])

16

In [9]:
df['Class'].value_counts()

2    458
4    241
Name: Class, dtype: int64

Since the Benign class "2" consist of 458, dropping 16 of those observations is okay.

In [10]:
# To replace the rows that contain "?" to Nan
df['Bare Nuclei'].replace("?", np.NAN, inplace=True)
df = df.dropna()

Note that for class: 2 is benign, 4 is for malignant

In [11]:
df['Bare Nuclei'].value_counts()

1     402
10    132
5      30
2      30
3      28
8      21
4      19
9       9
7       8
6       4
Name: Bare Nuclei, dtype: int64

In [12]:
# To convert the output feature "Class" to 0 and 1
df['Class'].value_counts()

2    444
4    239
Name: Class, dtype: int64

In [13]:
df['Class'] = (df['Class'] / 2) - 1

In [14]:
df['Class'].value_counts()

0.0    444
1.0    239
Name: Class, dtype: int64

Note that for class: 0 is benign, 1 is for malignant

In [15]:
df.columns

Index(['id', 'Clump Thickness', 'Uniformity of Cell Size',
       'Uniformity of Cell Shape', 'Marginal Adhesion',
       'Single Epithelial Cell Size', 'Bare Nuclei', 'Bland Chromatin',
       'Normal Nucleoli', 'Mitoses', 'Class'],
      dtype='object')

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 683 entries, 0 to 698
Data columns (total 11 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   id                           683 non-null    int64  
 1   Clump Thickness              683 non-null    int64  
 2   Uniformity of Cell Size      683 non-null    int64  
 3   Uniformity of Cell Shape     683 non-null    int64  
 4   Marginal Adhesion            683 non-null    int64  
 5   Single Epithelial Cell Size  683 non-null    int64  
 6   Bare Nuclei                  683 non-null    object 
 7   Bland Chromatin              683 non-null    int64  
 8   Normal Nucleoli              683 non-null    int64  
 9   Mitoses                      683 non-null    int64  
 10  Class                        683 non-null    float64
dtypes: float64(1), int64(9), object(1)
memory usage: 64.0+ KB


In [17]:
X = df.drop(['id', 'Class'], axis=1)
X_col = X.columns

In [18]:
y = df['Class']

### Scale the Data

In [19]:
from sklearn.preprocessing import MinMaxScaler

In [20]:
# Drop ID and Y variable
df1 = pd.DataFrame(df.drop(['id', 'Class'], axis=1), columns=X_col)

In [21]:
df1.head()

Clump Thickness  Uniformity of Cell Size  Uniformity of Cell Shape  \
0                5                        1                         1   
1                5                        4                         4   
2                3                        1                         1   
3                6                        8                         8   
4                4                        1                         1   

   Marginal Adhesion  Single Epithelial Cell Size Bare Nuclei  \
0                  1                            2           1   
1                  5                            7          10   
2                  1                            2           2   
3                  1                            3           4   
4                  3                            2           1   

   Bland Chromatin  Normal Nucleoli  Mitoses  
0                3                1        1  
1                3                2        1  
2                3                1        1  
3                3                7        1  
4                3                1        1

In [22]:
scaler = MinMaxScaler()

In [23]:
df1 = pd.DataFrame(scaler.fit_transform(df1), columns=X_col)

In [24]:
df1.head()

Clump Thickness  Uniformity of Cell Size  Uniformity of Cell Shape  \
0         0.444444                 0.000000                  0.000000   
1         0.444444                 0.333333                  0.333333   
2         0.222222                 0.000000                  0.000000   
3         0.555556                 0.777778                  0.777778   
4         0.333333                 0.000000                  0.000000   

   Marginal Adhesion  Single Epithelial Cell Size  Bare Nuclei  \
0           0.000000                     0.111111     0.000000   
1           0.444444                     0.666667     1.000000   
2           0.000000                     0.111111     0.111111   
3           0.000000                     0.222222     0.333333   
4           0.222222                     0.111111     0.000000   

   Bland Chromatin  Normal Nucleoli  Mitoses  
0         0.222222         0.000000      0.0  
1         0.222222         0.111111      0.0  
2         0.222222         0.000000      0.0  
3         0.222222         0.666667      0.0  
4         0.222222         0.000000      0.0

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
X_train, X_test, y_train, y_test = train_test_split(df1, y,
                                                    train_size=0.8,
                                                    random_state=42)

## Model

In [27]:
from sklearn.linear_model import SGDClassifier

In [28]:
from sklearn import preprocessing
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.metrics import accuracy_score, classification_report
from sklearn.metrics import confusion_matrix, roc_auc_score
def print_score(clf, X_train, X_test, y_train, y_test, train=True):
    '''
    v0.1 Follow the scikit learn library format in terms of input
    print the accuracy score, classification report and confusion matrix of classifier
    '''
    lb = preprocessing.LabelBinarizer()
    lb.fit(y_train)
    if train:
        '''
        training performance
        '''
        res = clf.predict(X_train)
        print("Train Result:\n")
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_train, 
                                                                res)))
        print("Classification Report: \n {}\n".format(classification_report(y_train, 
                                                                            res)))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_train, 
                                                                  res)))
        print("ROC AUC: {0:.4f}\n".format(roc_auc_score(lb.transform(y_train), 
                                                      lb.transform(res))))

        res = cross_val_score(clf, X_train, y_train, cv=10, scoring='accuracy')
        print("Average Accuracy: \t {0:.4f}".format(np.mean(res)))
        print("Accuracy SD: \t\t {0:.4f}".format(np.std(res)))
        
    elif train==False:
        '''
        test performance
        '''
        res_test = clf.predict(X_test)
        print("Test Result:\n")        
        print("accuracy score: {0:.4f}\n".format(accuracy_score(y_test, 
                                                                res_test)))
        print("Classification Report: \n {}\n".format(classification_report(y_test, 
                                                                            res_test)))
        print("Confusion Matrix: \n {}\n".format(confusion_matrix(y_test, 
                                                                  res_test)))   
        print("ROC AUC: {0:.4f}\n".format(roc_auc_score(lb.transform(y_test), 
                                                      lb.transform(res_test))))

In [29]:
SGDmodel = SGDClassifier()

In [45]:
SGDmodel.fit(X_train, y_train)

SGDClassifier()

In [46]:
print_score(SGDmodel, X_train, X_test, y_train, y_test, train=True)

Train Result:

accuracy score: 0.9689

Classification Report: 
               precision    recall  f1-score   support

         0.0       1.00      0.96      0.98       365
         1.0       0.92      0.99      0.95       181

    accuracy                           0.97       546
   macro avg       0.96      0.98      0.97       546
weighted avg       0.97      0.97      0.97       546


Confusion Matrix: 
 [[349  16]
 [  1 180]]

ROC AUC: 0.9753

Average Accuracy: 	 0.9579
Accuracy SD: 		 0.0201


In [47]:
print_score(SGDmodel, X_train, X_test, y_train, y_test, train=False)

Test Result:

accuracy score: 0.9927

Classification Report: 
               precision    recall  f1-score   support

         0.0       1.00      0.99      0.99        79
         1.0       0.98      1.00      0.99        58

    accuracy                           0.99       137
   macro avg       0.99      0.99      0.99       137
weighted avg       0.99      0.99      0.99       137


Confusion Matrix: 
 [[78  1]
 [ 0 58]]

ROC AUC: 0.9937



### Save the model

In [48]:
import pickle

In [49]:
with open("model", "wb") as file:
    pickle.dump(SGDmodel, file)

### Save the scalar

In [50]:
with open("scaler", "wb") as file:
    pickle.dump(SGDmodel, file)